include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+};
+
+/*!\brief A single-use synchronisation point for closable concurrent data structures.
+ *
+ * \details
+ *
+ * A reader writer latch is a thread coordination mechanism specifically designed to synchronise with a concurrent
+ * data structure such as the seqan3::contrib::buffer_queue. In particular, the latch is constructed with a
+ * seqan3::contrib::reader_count and a seqan3::contrib::writer_count and offers special functions for readers that
+ * arrive and writers that arrive at the synchronisation point. If all writers arrived at the synchronisation point
+ * the associated concurrent data structure is closed in a completion phase by one of the participating producer
+ * threads. Thus, the concurrent data structure will be automatically signalled that there is no more data coming.
+ * If all participating readers arrived at the synchronisation point no completion phase is invoked, but the
+ * destructor of the reader_writer_latch will wait until all readers, respectively writers arrived at the
+ * synchronisation point. This allows to protect a concurrent data structure against an early termination.
+ * One can use the seqan3::contrib::scoped_reader_guard and seqan3::contrib::scoped_writer_guard to associate a thread
+ * with the reader_writer_latch as either a reaader or writer.
+ *
+ * The reader_writer_latch class is non-copyable and not default constructible.
+ */
+class reader_writer_latch
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    reader_writer_latch()                                        = delete;  //!< Deleted.
+    reader_writer_latch(reader_writer_latch const &)             = delete;  //!< Deleted.
+    reader_writer_latch(reader_writer_latch &&)                  = default; //!< Defaulted.
+    reader_writer_latch & operator=(reader_writer_latch const &) = delete;  //!< Deleted.
+    reader_writer_latch & operator=(reader_writer_latch &&)      = default; //!< Defaulted.
+    ~reader_writer_latch()                                       = default; //!< Defaulted.
+
+    /*!\brief Constructs the reader_writer_latch with the reader count, writer count and the associated data structure.
+     * \tparam concurrent_t The type of the concurrent data structure. Must have a member function `close`.
+     * \param rcount The expected number of consumer threads.
+     * \param wcount The expected number of producer threads.
+     * \param ds     The concurrent data structure.
+     *
+     * \details
+     *
+     * Initialises a std::contrib::latch for the consumer threads and a std::contrib::latch for the producer threads.
+     * Initialises the completion function which calls `ds.close()` on completion. Note, only a reference to the
+     * concurrent data structure is stored. The caller must ensure that `ds` is not destructed before this.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable queue.
```

Comment:
closable queue or closable concurrent data structure?

Reply:
It should be generalised, because I also need to have a concurrent stack interface for the bgzf_outstream

Reply:
I meant the comment. I think the comment should be `closable concurrent data structure`